core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java

```diff
@@ -37,6 +37,7 @@
 import io.trino.spi.classloader.ThreadContextClassLoader;
```

Comment:
When a connector says "I use global roles", the role methods are never called on the connector.  The connector security system can inspect the "enabled system roles" in the identity for access control checks it it wants.  This is what the file based connector access control in this PR does.

Reply:
> When a connector says "I use global roles", the role methods are never called on the connector. 

When a connector says "I use system roles" then role methods are never called on the system, right?

It is incoherent with what we have now with everything related to access control. We were always doing full cascade of checks. In order to have an access, system would grant access and catalog would need to grant access. So goes to row filters and column masks we capture them from both access control levels.

I think user need to has an option. I he configures both levels then they expect to use both levels of access control. Like using file based system access control together with Hive Sql standard today. This change will break such use cases.

Reply:
> > When a connector says "I use global roles", the role methods are never called on the connector.

> When a connector says "I use system roles" then role methods are never called on the system, right?

No.  What I wrote is want I meant.  When a connector does not have it's own roles, then roles are handles at the system level, so the catalog isn't consulted about things like set role.

> I think user need to has an option. I he configures both levels then they expect to use both levels of access control. Like using file based system access control together with Hive Sql standard today. This change will break such use cases.

I tried implementing that.  It was insanely complex, and it was really confusing to reason about.   This is the third (maybe forth) rewrite of this feature.

Reply:
> When a connector says "I use system roles" then role methods are never called on the system, right?

🤦 I wanted to say: _then role methods are never called on the connector_.

> I tried implementing that. It was insanely complex, and it was really confusing to reason about.

It is one time effort. If we go with your approach we will need live with this for years, so if design is bad, misleading or complex to users, plugin developers or Trino contributors then I think we are going to pay a bigger price for this in a long run.

Why it was that complex? Maybe we can find another way to solve this? The fact that system and connector access controls are independent and evaluated in a cascade way is simple. Changing this with introducing a bridge that sometimes connector is using system roles breaks that design in my opinion.

Is it all about the syntax? That it may be ambiguous? If it is a case then let's solve it at syntax layer. Making syntax clear has an advantage that it will be clear to everyone: users, contributors and design.

For example if you are introducing system roles then maybe we could have, one of these:
 -  `SYSTEM GRANT role TO ...`, `SYSTEM GRANT SELECT ...`
 - `GRANT SYSTEM role ...`, `GRANT SYSTEM SELECT...`
 - `GRANT role ... IN SYSTEM`, `GRANT SELECT ... IN SYSTEM`

We could replace `SYSTEM` with `GLOBAL`. Any referred role would be a system role.

Alternatively, if we believe that system roles are first class citizen and connector roles are second class, then we could change syntax only for connector roles to always require `IN catalog` to be always specified.